File: 11_container_with_most_water/max_area.go
package maxarea

func maxArea(height []int) int {
	max := 0

	l, r := 0, len(height)-1

	for l < r {
		lHeight, rHeight := height[l], height[r]
		square := min(lHeight, rHeight) * (r - l)

		if square > max {
			max = square
		}

		if lHeight < rHeight {
			l++
		} else {
			r--
		}
	}

	return max
}

func min(first, second int) int {
	if first < second {
		return first
	}

	return second
}

File: 11_container_with_most_water/MaxArea.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int maximum = 0;

        int l = 0;
        int r = height.size() - 1;

        while (l < r) {
            int lHeight = height[l];
            int rHeight = height[r];

            int square = min(lHeight, rHeight) * (r - l);

            if (square > maximum) {
                maximum = square;
            }

            if (lHeight < rHeight) {
                l++;
            } else {
                r--;
            }
        }

        return maximum;
    }
};